dChatServer/ChatPacketHandler.cpp

```diff
@@ -337,121 +337,142 @@ void ChatPacketHandler::HandleRemoveFriend(Packet* packet) {
+	LUWString message(size);
+	inStream.Read(message);
+	
+	LOG("Got a message from (%s) via [%s]: %s", sender.playerName.c_str(), StringifiedEnum::ToString(channel).data(), message.GetAsString().c_str());
 
-	if (channel != 8) return;
+	if (channel == eChatChannel::TEAM) {
+		auto* team = Game::playerContainer.GetTeam(playerID);
+		if (team == nullptr) return;
 
-	auto* team = Game::playerContainer.GetTeam(playerID);
+		for (const auto memberId : team->memberIDs) {
+			const auto& otherMember = Game::playerContainer.GetPlayerData(memberId);
 
-	if (team == nullptr) return;
+			if (!otherMember) return;
+
+			CBITSTREAM;
+			BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT_INTERNAL, eChatInternalMessageType::ROUTE_TO_PLAYER);
+			bitStream.Write(otherMember.playerID);
+
+			BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT, eChatMessageType::PRIVATE_CHAT_MESSAGE);
+			bitStream.Write(otherMember.playerID);
+			bitStream.Write(eChatChannel::TEAM);
+			bitStream.Write<uint32_t>(0); // not used
+			bitStream.Write(LUWString(sender.playerName));
+			bitStream.Write(sender.playerID);
+			bitStream.Write<uint16_t>(0); // sourceID
+			bitStream.Write(sender.GMLevel);
+			bitStream.Write(LUWString(otherMember.playerName));
+			bitStream.Write(otherMember.GMLevel);
+			bitStream.Write(eChatMessageResponseCode::Sent);
+			bitStream.Write(message);
+
+			SystemAddress sysAddr = otherMember.sysAddr;
+			SEND_PACKET;
+		}
+	}
+}
 
-	for (const auto memberId : team->memberIDs) {
-		const auto& otherMember = Game::playerContainer.GetPlayerData(memberId);
+void ChatPacketHandler::HandlePrivateChatMessage(Packet* packet) {
+	CINSTREAM_SKIP_HEADER;
+	LWOOBJID playerID;
+	inStream.Read(playerID);
 
-		if (!otherMember) return;
+	const auto& sender = Game::playerContainer.GetPlayerData(playerID);
+	if (!sender || sender.GetIsMuted()) return;
 
-		CBITSTREAM;
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT_INTERNAL, eChatInternalMessageType::ROUTE_TO_PLAYER);
-		bitStream.Write(otherMember.playerID);
+	eChatChannel channel;
+	uint32_t size;
+	LUWString receiverName;
 
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT, eChatMessageType::PRIVATE_CHAT_MESSAGE);
-		bitStream.Write(otherMember.playerID);
-		bitStream.Write<uint8_t>(8);
-		bitStream.Write<unsigned int>(69);
-		bitStream.Write(LUWString(sender.playerName));
-		bitStream.Write(sender.playerID);
-		bitStream.Write<uint16_t>(0);
-		bitStream.Write<uint8_t>(0); //not mythran nametag
-		bitStream.Write(LUWString(otherMember.playerName));
-		bitStream.Write<uint8_t>(0); //not mythran for receiver
-		bitStream.Write<uint8_t>(0); //teams?
-		bitStream.Write(LUWString(message, 512));
-
-		SystemAddress sysAddr = otherMember.sysAddr;
-		SEND_PACKET;
-	}
-}
+	inStream.IgnoreBytes(4);
+	inStream.Read(channel);
+	inStream.Read(size);
+	inStream.IgnoreBytes(77);
+	inStream.Read(receiverName);
+	inStream.IgnoreBytes(2);
```

Comment:
Same thing here with the magic numbers.  Try to elaborate on them if possible.

Reply:
> Same thing here with the magic numbers. Try to elaborate on them if possible.

If this code is also repeated, consider a function to deserialize and a struct to propagate the data between functions

Reply:
yeah, was thinking about breaking it out and how to do that

Reply:
erh, not the reading, but the writing.
the reading is just slightly different enough to be really annoying to consolidate to a common place like that

Reply:
will consolidate the writing in a future PR